File: Solution/CSharp/141.LinkedListCycle.cs
public class Solution {
    public bool HasCycle(ListNode head) {
        if (head == null) {
            return false;
        }
        ListNode node = head;
        HashSet<int> hashSet = new HashSet<int>();
        hashSet.Add(node.GetHashCode());
        while (node.next != null) {
            node = node.next;
            int hash = node.GetHashCode();
            if (hashSet.Contains(hash)) {
                return true;
            } else {
                hashSet.Add(hash);
            }
        }
        return false;
    }
}
public class Solution {
    public bool HasCycle(ListNode head) {
        if (head == null || head.next == null) {
            return false;
        }
        ListNode slow = head;
        ListNode fast = head;
        while (fast.next != null && fast.next.next != null) {
            slow = slow.next;
            fast = fast.next.next;
            if (slow == fast) {
                return true;
            }
        }
        return false;
    }
}

File: Solution/Python/141. Linked List Cycle.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution(object):
    def hasCycle(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        if not head:
            return False
        slow = head
        fast = head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if slow == fast:
                return True
        return False


# 1>2>3>
head = ListNode(1)
head.next = ListNode(2)
head.next.next = ListNode(3)
head.next.next.next = head
res = Solution().hasCycle(head)

print(res)
